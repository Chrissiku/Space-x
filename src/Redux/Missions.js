const BASE_URL = 'https://api.spacexdata.com/v3/missions';
const LOAD_MISSIONS = 'missions/load';
const JOIN_MISSION = 'missions/join';
const LEAVE_MISSION = 'missions/leave';

const loadMissions = (payload) => ({
  type: LOAD_MISSIONS,
  payload,
});

export const joinMission = (id) => ({
  type: JOIN_MISSION,
  id,
});

export const leaveMission = (id) => ({
  type: LEAVE_MISSION,
  id,
});

export const fetchMissions = async (dispatch) => {
  const response = await fetch(BASE_URL);
  const missions = await response.json();

  dispatch(
    loadMissions(
      missions.map((mission) => ({
        id: mission.mission_id,
        name: mission.mission_name,
        description: mission.description,
      })),
    ),
  );
};

const MissionsReducer = (state = [], action) => {
  switch (action.type) {
    case LOAD_MISSIONS:
      return action.payload;
    case JOIN_MISSION:
      return state.map((mission) => {
        if (mission.id !== action.id) {
          return mission;
        }

        return {
          ...mission,
          reserved: true,
        };
      });
    case LEAVE_MISSION:
      return state.map((mission) => {
        if (mission.id !== action.id) {
          return mission;
        }

        return {
          ...mission,
          reserved: false,
        };
      });
    default:
      return state;
  }
};

export default MissionsReducer;
