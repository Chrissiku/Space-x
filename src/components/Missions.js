const Mission = () => <h1>Missions Page</h1>;

export default Mission;
