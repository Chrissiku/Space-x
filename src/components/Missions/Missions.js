/* eslint-disable linebreak-style */
/* eslint-disable no-unused-vars */
/* eslint-disable quotes */
import { useEffect } from "react";
import Container from "react-bootstrap/Container";
import Table from "react-bootstrap/Table";
import Badge from "react-bootstrap/Badge";
import Button from "react-bootstrap/Button";
import { useSelector, useDispatch } from "react-redux";
import { fetchMissions, joinMission, leaveMission } from "../../Redux/Missions";

const Missions = () => {
  const dispatch = useDispatch();
  const missions = useSelector((state) => state.missions);

  useEffect(() => {
    if (!missions.length) {
      dispatch(fetchMissions);
    }
  }, []);

  const joinMissionHandler = (id) => dispatch(joinMission(id));
  const leaveMissionHandler = (id) => dispatch(leaveMission(id));

  return (
    <Container>
      <Table striped bordered hover>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th>{"  "}</th>
          </tr>
        </thead>
        <tbody>
          {missions.map(({
            id, name, description, reserved,
          }) => (
            <tr key={id}>
              <td>
                <p className="fw-bold">{name}</p>
              </td>
              <td>
                <p>{description}</p>
              </td>
              <td className="align-middle">
                {reserved && <Badge bg="success">Active Member</Badge>}
                {!reserved && <Badge bg="secondary">NOT A MEMBER</Badge>}
              </td>
              <td className="col-2 align-middle text-center">
                {reserved && (
                <Button
                  size="sm"
                  variant="outline-danger"
                  onClick={() => leaveMissionHandler(id)}
                >
                  Leave Mission
                </Button>
                )}
                {!reserved && (
                <Button
                  size="sm"
                  variant="outline-secondary"
                  onClick={() => joinMissionHandler(id)}
                >
                  Join Mission
                </Button>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </Table>
    </Container>
  );
};

export default Missions;
