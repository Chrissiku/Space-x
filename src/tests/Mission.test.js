/* eslint-disable linebreak-style */
/* eslint-disable comma-dangle */
/* eslint-disable quotes */
import React from "react";
import TestRenderer from "react-test-renderer";
import { Provider } from "react-redux";
import store from "../Redux/ConfigureStore";
import Missions from "../components/Missions/Missions";

it("Missions render correctly", () => {
  const Tree = TestRenderer.create(
    <Provider store={store}>
      <Missions />
    </Provider>
  );
  expect(Tree).toMatchSnapshot();
});
