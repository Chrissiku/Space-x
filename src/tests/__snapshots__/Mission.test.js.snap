// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions render correctly 1`] = `
<div
  className="container"
>
  <table
    className="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th>
            
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
